sourcepolicy/pb/policy.proto

```diff
@@ -0,0 +1,62 @@
+syntax = "proto3";
+
+package moby.buildkit.v1.sourcepolicy;
+
+// Rule defines the action(s) to take whe.n a source is matched
+message Rule {
+	PolicyAction action = 1; 
+	Source source = 2;
+	// Destination is an optional value that can be set for certain actions (e.g. `CONVERT`)
+	Destination destination = 3;
+}
+
+// DestinationIdentifier is used for action types that involves mutating the source
+message Destination {
+	string identifier = 1;
+	map<string, string> attrs = 2;
+}
+
+// SourceIdentifier identifies a source to match a policy to
+message Source {
+	string type = 1;
+	string identifier = 2;
+	// MatchType is the type of match to perform on the source identifier
+	MatchType match_type = 3;
+	repeated AttrConstraint constraints = 4;
+}
+
+// PolicyAction defines the action to take when a source is matched
+enum PolicyAction {
+	ALLOW = 0;
+	DENY = 1;
+	CONVERT = 2;
+}
+
+// AttrConstraint defines a constraint on a source attribute
+message AttrConstraint {
+	string key = 1;
+	string value = 2;
+	AttrMatch condition = 3;
+}
+
+// AttrMatch defines the condition to match a source attribute
+enum AttrMatch {
+	EQUAL = 0;
+	NOTEQUAL = 1;
+	MATCHES = 2;
+}
+
+// Policy is the list of rules the policy engine will perform
+message Policy {
+	int64 version = 1; // Currently 1
+	repeated Rule rules = 2;
+}
+
+// Match type is used to determine how a rule source is matched
+enum MatchType {
+	// EXACT treats the soruce identifier as a litteral string match
+	EXACT = 0;
+	// REGEX treats the source identifier as a regular expression
+	// With regex matching you can also use match groups to replace values in the destination identifier
+	REGEX = 1;
+}
```

Comment:
I thought the idea was to use wildcards. Regex is useful too, though

Reply:
<details>
<summary><code>util/wildcard/wildcard.go</code></summary>

<p>

```go
package wildcard

import (
        "errors"
        "fmt"
        "regexp"
        "strings"
)

// New returns a wildcard object for a string that contains "*" symbols.
func New(s string) (*Wildcard, error) {
        reStr, err := wildcard2regexp(s)
        if err != nil {
                return nil, fmt.Errorf("failed to translate wildcard %q to regexp: %w", s, err)
        }
        re, err := regexp.Compile(reStr)
        if err != nil {
                return nil, fmt.Errorf("failed to compile regexp %q (translated from wildcard %q): %w", reStr, s, err)
        }
        w := &Wildcard{
                orig: s,
                re:   re,
        }
        return w, nil
}

func wildcard2regexp(wildcard string) (string, error) {
        s := regexp.QuoteMeta(wildcard)
        if strings.Contains(s, "\\*\\*") {
                return "", errors.New("invalid wildcard: \"**\"")
        }
        s = strings.ReplaceAll(s, "\\*", "(.*)")
        s = "^" + s + "$"
        return s, nil
}

// Wildcard is a wildcard matcher object.
type Wildcard struct {
        orig string
        re   *regexp.Regexp
}

// String implements fmt.Stringer.
func (w *Wildcard) String() string {
        return w.orig
}

// Match returns a non-nil Match on match.
func (w *Wildcard) Match(q string) *Match {
        submatches := w.re.FindStringSubmatch(q)
        if len(submatches) == 0 {
                return nil
        }
        m := &Match{
                w:          w,
                Submatches: submatches,
                // FIXME: avoid executing regexp twice
                idx: w.re.FindStringSubmatchIndex(q),
        }
        return m
}

// Match is a matched result.
type Match struct {
        w          *Wildcard
        Submatches []string // 0: the entire query, 1: the first submatch, 2: the second submatch, ...
        idx        []int
}

// String implements fmt.Stringer.
func (m *Match) String() string {
        if len(m.Submatches) == 0 {
                return ""
        }
        return m.Submatches[0]
}

// Format formats submatch strings like "$1", "$2".
func (m *Match) Format(f string) (string, error) {
        if m.w == nil || len(m.Submatches) == 0 || len(m.idx) == 0 {
                return "", errors.New("invalid state")
        }
        var b []byte
        b = m.w.re.ExpandString(b, f, m.Submatches[0], m.idx)
        return string(b), nil
}
```

</p>

</details>

<details>
<summary><code>util/wildcard/wildcard_test.go</code></summary>

<p>

```go
package wildcard

import (
        "testing"

        "github.com/stretchr/testify/assert"
)

func TestWildcard(t *testing.T) {
        wildcardStr := "docker.io/*/alpine:*"
        wildcard, err := New(wildcardStr)
        assert.NoError(t, err)
        t.Run("Match", func(t *testing.T) {
                m := wildcard.Match("docker.io/library/alpine:latest")
                assert.Equal(t, []string{"docker.io/library/alpine:latest", "library", "latest"}, m.Submatches)
                s, err := m.Format("$1-${2}-$3-$$-$0")
                assert.NoError(t, err)
                // "$3" is replaced with an empty string without producing an error, because Format() internally uses regexp.*Regexp.Expand():
                // https://pkg.go.dev/regexp#Regexp.Expand
                assert.Equal(t, "library-latest--$-docker.io/library/alpine:latest", s)
        })
        t.Run("NoMatch", func(t *testing.T) {
                assert.Nil(t, wildcard.Match("docker.io/library/busybox:latest"))
                assert.Nil(t, wildcard.Match("alpine:latest"), "matcher must not be aware of the Docker Hub reference convention")
        })
}

func TestWildcardInvalid(t *testing.T) {
        wildcardStr := "docker.io/library/alpine:**"
        _, err := New(wildcardStr)
        assert.ErrorContains(t, err, "invalid wildcard: \"**\"")
}

func TestWildcardEscape(t *testing.T) {
        wildcardStr := "docker.io/library/alpine:\\*"
        wildcard, err := New(wildcardStr)
        assert.NoError(t, err)
        t.Run("NoMatch", func(t *testing.T) {
                assert.Nil(t, wildcard.Match("docker.io/library/alpine:latest"))
        })
}

func TestWildcardParentheses(t *testing.T) {
        wildcardStr := "docker.io/library/alpine:(*)"
        wildcard, err := New(wildcardStr)
        assert.NoError(t, err)
        t.Run("NoMatch", func(t *testing.T) {
                assert.Nil(t, wildcard.Match("docker.io/library/alpine:latest"))
        })
}
```

</p>

</details>

Reply:
Oooh good to know that package exists, didn't realize it was a thing when I was hunting down pre-existing work to deal with wildcards.

Reply:
Not really "exists", just copied from my local abandoned and unpushed branch 😅

Reply:
Added this and put you as co-author on the commit.